### DISCUSSION TITLE
Custom libraries per user

### DISCUSSION THREAD

**Discussion by amducious** [navidrome/navidrome#2452] [Ideas]
Right now this software is great if you are the only person using it, or you have similar taste in music as the other users. 

What would be great, is to make this software work more like Spotify/iTunes/Deezer. In the way that each user that logs in has their own private library that they can curate from the "Back catalog". The reasoning behind this is I have multiple users who want to use this service/server but we have drastically different taste in music. and truthfully I'm sick of having kids music in my Library lol. 

The idea
--------------------
1. The current "Library" become a back catalog that each user can search and/or browse. 
2. Each user that is created gets their own "private library" that they can pull music in from the "back catalog"
3. Each user that is created has the option to either 
    a. pull the entire "back catalog" into their library
    b. have a blank library created witch they choose to fill with the music they want.

**Comment by dweymouth** [comment]
Multiple library support is on the roadmap but I think the plan there was to be able to assign specific music folders to different users, not have a dynamic library that users could choose from the "back catalog". What you're describing would be a cool feature but for now it can kind of be approximated by marking albums as favorites. Each user could then browse their favorite albums as their main portal, with the ability to search and play from the full library as well. It's not perfect since you can't apply sorting orders to your favorite albums (e.g. recently added, random), and you can't in most Subsonic clients either since the API only exposes sort orders on the main collection.

**Comment by amducious** [comment]
I was thinking the same thing with the favorites work around. But favoriting things did not work the way I thought it would. 
1. If I favorite an album then it is listed in the "favorite" tab. But when I go to artists and filter by favorite that artists of that album is not listed. 
2. if I go to artist and mark one as a favorite, that artists albums do not show in the favorites tab. 
3. If I favorite a song, same thing as above it dose not show up as a favorite artists or in the Favorites Tab. 

Is this a bug, or is this the way it "should" be working?

**Reply by dweymouth** [reply]
That is how it "should" be working, since the feature really isn't a "custom library" but a way to highlight individual artists, albums, and songs that you want quick access to. Since artists, albums, and songs can all be favorited independently it makes sense that the behavior is as it is.

**Comment by hashhar** [comment]
I researched this in depth and came to a very sad conclusion. The way things work now is mostly because of how Subsonic API itself works. Subsonic was built around the concept of folders as libraries and it's pretty ingrained (there are APIs that return file paths).

Otherwise the proper way would be to have a `user_library` table with every user's library entities in there and then have a view in the UI to only browse this "library" instead of the whole catalog. Searches etc. can also be scoped to the library by filtering on the active user this way.

The mechanic itself isn't hard to implement but it's impossible to make it work on the player side sadly. Or I'm not smart enough.

**Reply by deluan** [reply]
Hey, not sure what do you mean. What you are describing was implemented July last year, in version [0.58.0](https://github.com/navidrome/navidrome/releases/tag/v0.58.0), and is documented here: https://www.navidrome.org/docs/usage/features/multi-library/

The Subsonic API provides multi-library support for (most) its endpoints, see the param `musicFolderId`.

Is there anything missing in Navidrome implementation?

**Reply by hashhar** [reply]
hah, you also had the idea to overload the purpose of `musicFolderId`? See https://github.com/opensubsonic/open-subsonic-api/discussions/229 for an example of what I was imagining (and the discarded idea of using `musicFolderId` at "Virtual musicFolderId workaround" in that discussion). Happy to hear from you since Navidrome will likely be the server where I'd implement the first cut for the spec.

**Reply by deluan** [reply]
Navidrome does not "overload" its purpose: the `musicFolderId`, in Subsonic terms, _is_ the `libraryId` in Navidrome, so this already works as expected. If other Subsonic servers do not implement it, that's on them to comply to the API. The spec is already there. We just need to extend it to a few endpoints (like getPlaylist/getPlaylists). See https://github.com/opensubsonic/open-subsonic-api/pull/162

**Reply by hashhar** [reply]
Thanks for the link to the other proposal. Multi-library is great for the admin-managed case. But it solves a different problem than what I'm after.

In multi-library the admin pre-organizes files into directories, creates libraries from those paths and assigns users to them. The structure is static and decided at the filesystem level.
What I'm proposing is user-managed curation. A user with access to 100k tracks (across music folders/libraries or in a single folder/library) wants to say "these 30k tracks are *my* library" dynamically from within a client regardless of where the files live on disk. Think of "add to library" like in streaming services. The admin provides a catalog and users can curate their own sub-sections of the catalog. Right now the only solution to this is to have a playlist.

In my case my family shares one Navidrome instance with one big music folder. With multi-library, I need to physically split the files into `/music/me/`, `/music/kids/`, etc. and overlapping taste means duplicating files or giving up. With a user library, everyone sees the same catalog but each person marks what's theirs and shuffle/browse/search only shows what each person cares about.

The two features compose well IMO, `musicFolderId=2&inLibrary=true` would mean "from my lossless library, only stuff I've added to my collection." Multi-library scopes what you *can* see, user library scopes what you *want* to see.

The full proposal is at opensubsonic/open-subsonic-api#229 - I'd love your input.
